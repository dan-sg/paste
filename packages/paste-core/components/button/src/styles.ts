import {css, styled, themeGet} from '@twilio-paste/styling-library';
import {Absolute} from '@twilio-paste/absolute';
import {ThemeShape} from '@twilio-paste/theme';
import {ButtonWrapperProps, ButtonChildrenProps} from './types';

/*
 * Sizes
 */
const sizeReset = {
  fontSize: '100%',
  padding: 'space0',
};
const sizeIcon = {
  padding: 'space30',
  borderRadius: 'borderRadius20',
  fontSize: '100%',
  /* To fix abnormal button padding-bottom */
  lineHeight: 'unset',
};
const sizeSmall = {
  paddingTop: 'space10',
  paddingBottom: 'space10',
  paddingRight: 'space30',
  paddingLeft: 'space30',
  borderRadius: 'borderRadius10',
  fontSize: 'fontSize30',
  lineHeight: 'lineHeight30',
};
const sizeDefault = {
  paddingTop: 'space30',
  paddingBottom: 'space30',
  paddingRight: 'space50',
  paddingLeft: 'space50',
  borderRadius: 'borderRadius20',
  fontSize: 'fontSize30',
  lineHeight: 'lineHeight30',
};

/*
 * Base
 */
const baseButtonWrapper = {
  /* Hide default browser styles */
  appearance: 'none',
  border: 'none',
  display: 'inline-block',
  outline: 'none',
  background: 'none',
  transition: 'background-color 100ms ease-in, box-shadow 100ms ease-in',
  fontFamily: 'fontFamilyText',
  fontWeight: 'fontWeightSemibold',
  /* Remove extra black dotted border FF adds */
  '&::-moz-focus-inner': {
    border: 'none',
  },
};
const baseEnabled = {
  ...baseButtonWrapper,
  textDecoration: 'none',
  cursor: 'pointer',
  '&:focus, &:active': {
    boxShadow: 'shadowFocus',
  },
  /*
   defensively resetting from over zealous legacy global
   styles "a {...}" when button is set as an anchor
   */
  '&:hover, &:focus, &:active': {
    textDecoration: 'none',
  },
};
const baseLoading = {
  ...baseButtonWrapper,
  cursor: 'wait',
};
const baseDisabled = {
  ...baseButtonWrapper,
  cursor: 'not-allowed',
};

/*
 * Variants
 */
// Primary
const variantPrimaryBase = {
  color: 'colorTextInverse',
  /*
    defensively resetting from over zealous legacy global
    styles "a {...}" when button is set as an anchor
  */
  '&:hover,&:focus,&:active': {
    color: 'colorTextInverse',
  },
};
const variantPrimaryEnabled = {
  ...baseEnabled,
  ...variantPrimaryBase,
  backgroundColor: 'colorBackgroundPrimary',
  boxShadow: (props: ThemeShape) => `inset 0 0 0 ${props.space.space10}
      ${props.backgroundColors.colorBackgroundPrimary}`,
  '&:hover': {
    backgroundColor: 'colorBackgroundPrimaryDarker',
    boxShadow: (props: ThemeShape) => `inset 0 0 0 ${props.space.space10}
        ${props.backgroundColors.colorBackgroundPrimaryDarker}`,
  },
  '&:focus': {
    backgroundColor: 'colorBackgroundPrimary',
    boxShadow: (props: ThemeShape) =>
      `inset 0 0 0 ${props.space.space10} ${props.backgroundColors.colorBackgroundPrimaryDarker}, ${props.shadows.shadowFocus}`,
  },
  '&:active': {
    backgroundColor: 'colorBackgroundPrimaryDark',
    boxShadow: (props: ThemeShape) =>
      `inset 0 0 0 ${props.space.space10} ${props.backgroundColors.colorBackgroundPrimaryDarker}, ${props.shadows.shadowFocus}`,
  },
};
const variantPrimaryLoading = {
  ...baseLoading,
  ...variantPrimaryBase,
  '&, &:hover, &:active, &:focus': {
    backgroundColor: 'colorBackgroundPrimaryDarker',
    boxShadow: (props: ThemeShape) =>
      `inset 0 0 0 ${props.space.space10} ${props.backgroundColors.colorBackgroundPrimaryDarker}`,
  },
};
const variantPrimaryDisabled = {
  ...baseDisabled,
  ...variantPrimaryBase,

  backgroundColor: 'colorBackgroundPrimaryLight',
  boxShadow: (props: ThemeShape) => `inset 0 0 0 ${props.space.space10}
      ${props.backgroundColors.colorBackgroundPrimaryLight}`,

  '&:hover,&:active': {
    backgroundColor: 'colorBackgroundPrimaryLight',
    boxShadow: (props: ThemeShape) => `inset 0 0 0 ${props.space.space10}
        ${props.backgroundColors.colorBackgroundPrimaryLight}`,
  },
};

// Secondary
const variantSecondaryBase = {
  backgroundColor: 'colorBackgroundBody',
};
const variantSecondaryEnabled = {
  ...baseEnabled,
  ...variantSecondaryBase,
  color: 'colorTextLink',
  backgroundColor: 'colorBackgroundBody',
  boxShadow: (props: ThemeShape) => `inset 0 0 0 ${props.space.space10} ${props.borderColors.colorBorderPrimary}`,

  '&:hover': {
    color: 'colorTextLinkDarker',
    backgroundColor: 'colorBackgroundPrimaryLightest',
    boxShadow: (props: ThemeShape) => `inset 0 0 0 ${props.space.space10}
        ${props.borderColors.colorBorderPrimaryDarker}`,
  },
  '&:focus': {
    color: 'colorTextLinkDarker',
    backgroundColor: 'colorBackgroundPrimaryLightest',
    boxShadow: (props: ThemeShape) => `inset 0 0 0 ${props.space.space10}
          ${props.borderColors.colorBorderPrimaryDarker},
        ${props.shadows.shadowFocus}`,
  },
  '&:active': {
    color: 'colorTextLinkDarker',
    backgroundColor: 'colorBackgroundPrimaryLighter',
    boxShadow: (props: ThemeShape) => `inset 0 0 0 ${props.space.space10}
          ${props.borderColors.colorBorderPrimaryDarker},
        ${props.shadows.shadowFocus}`,
  },
};
const variantSecondaryLoading = {
  ...baseLoading,
  ...variantSecondaryBase,
  color: 'colorTextLinkDarker',
  backgroundColor: 'colorBackgroundPrimaryLighter',
  boxShadow: (props: ThemeShape) =>
    `inset 0 0 0 ${props.space.space10} ${props.borderColors.colorBorderPrimaryLighter}`,
};
const variantSecondaryDisabled = {
  ...baseDisabled,
  ...variantSecondaryBase,
  color: 'colorTextLinkLight',
  boxShadow: (props: ThemeShape) => `inset 0 0 0 ${props.space.space10} ${props.borderColors.colorBorderPrimaryLight}`,
};

// Destructive
const variantDestructiveBase = {
  backgroundColor: 'colorBackgroundDestructive',
  color: 'colorTextInverse',
  boxShadow: (props: ThemeShape) =>
    `inset 0 0 0 ${props.space.space10} ${props.backgroundColors.colorBackgroundDestructive}`,
  /*
   * Console fixme
   * defensively resetting from over zealous legacy global
   * styles "a {...}" when button is set as an anchor
   */
  '&:hover, &:focus, &:active': {
    color: 'colorTextInverse',
  },
};
const variantDestructiveEnabled = {
  ...baseEnabled,
  ...variantDestructiveBase,
  '&:hover': {
    backgroundColor: 'colorBackgroundDestructiveDarker',
    boxShadow: (props: ThemeShape) =>
      `inset 0 0 0 ${props.space.space10} ${props.backgroundColors.colorBackgroundDestructiveDarker}`,
  },
  '&:focus': {
    backgroundColor: 'colorBackgroundDestructive',
    boxShadow: (props: ThemeShape) =>
      `inset 0 0 0 ${props.space.space10} ${props.backgroundColors.colorBackgroundDestructiveDarker}, ${props.shadows.shadowFocus}`,
  },
  '&:active': {
    backgroundColor: 'colorBackgroundDestructiveDark',
    boxShadow: (props: ThemeShape) =>
      `inset 0 0 0 ${props.space.space10} ${props.backgroundColors.colorBackgroundDestructiveDarker}, ${props.shadows.shadowFocus}`,
  },
};
const variantDestructiveLoading = {
  ...baseLoading,
  ...variantDestructiveBase,
  backgroundColor: 'colorBackgroundDestructiveDarker',
  boxShadow: (props: ThemeShape) =>
    `inset 0 0 0 ${props.space.space10} ${props.backgroundColors.colorBackgroundDestructiveDarker}`,
};
const variantDestructiveDisabled = {
  ...baseDisabled,
  ...variantDestructiveBase,
  backgroundColor: 'colorBackgroundDestructiveLight',
  boxShadow: (props: ThemeShape) =>
    `inset 0 0 0 ${props.space.space10} ${props.backgroundColors.colorBackgroundDestructiveLight}`,
};

// Destructive Link
const variantDestructiveLinkBase = {
  background: 'none',
};
const variantDestructiveLinkEnabled = {
  ...baseEnabled,
  ...variantDestructiveLinkBase,
  color: 'colorTextLinkDestructive',
  '&:hover, &:focus, &:active': {
    textDecoration: 'underline',
  },
  '&:hover': {
    color: 'colorTextLinkDestructiveDark',
  },
  '&:active': {
    color: 'colorTextLinkDestructiveDarker',
  },
};
const variantDestructiveLinkLoading = {
  ...baseLoading,
  ...variantDestructiveLinkBase,
  color: 'colorTextLinkDestructiveDarker',
};
const variantDestructiveLinkDisabled = {
  ...baseDisabled,
  ...variantDestructiveLinkBase,
  color: 'colorTextLinkDestructiveLight',
};

// Link
const variantLinkBase = {
  background: 'none',
};
const variantLinkEnabled = {
  ...baseEnabled,
  ...variantLinkBase,
  color: 'colorTextLink',
  '&:hover, &:focus, &:active': {
    textDecoration: 'underline',
  },
  '&:hover': {
    color: 'colorTextLinkDark',
  },
  '&:active': {
    color: 'colorTextLinkDarker',
  },
};
const variantLinkLoading = {
  ...baseLoading,
  ...variantLinkBase,
  color: 'colorTextLinkDarker',
};
const variantLinkDisabled = {
  ...baseDisabled,
  ...variantLinkBase,
  color: 'colorTextLinkLight',
};

// Reset
const variantResetEnabled = baseEnabled;
const variantResetLoading = baseLoading;
const variantResetDisabled = baseDisabled;

/*
 * Style getters
 */
const buttonSizeStyles = (props: ButtonWrapperProps): {} => {
  switch (props.size) {
    case 'reset':
      return sizeReset;
    case 'icon':
      return sizeIcon;
    case 'small':
      return sizeSmall;
    case 'default':
    default:
      return sizeDefault;
  }
};

const buttonStateVariantStyles = ({variant, buttonState}: ButtonWrapperProps): {} => {
  const isDisabled = buttonState === 'disabled';
  const isLoading = buttonState === 'loading';

  switch (variant) {
    case 'reset':
      if (isDisabled) return variantResetDisabled;
      if (isLoading) return variantResetLoading;
      return variantResetEnabled;
    case 'secondary':
      if (isDisabled) return variantSecondaryDisabled;
      if (isLoading) return variantSecondaryLoading;
      return variantSecondaryEnabled;
    case 'link':
      if (isDisabled) return variantLinkDisabled;
      if (isLoading) return variantLinkLoading;
      return variantLinkEnabled;
    case 'destructive':
      if (isDisabled) return variantDestructiveDisabled;
      if (isLoading) return variantDestructiveLoading;
      return variantDestructiveEnabled;
    case 'destructive_link':
      if (isDisabled) return variantDestructiveLinkDisabled;
      if (isLoading) return variantDestructiveLinkLoading;
      return variantDestructiveLinkEnabled;
    case 'primary':
    default:
      if (isDisabled) return variantPrimaryDisabled;
      if (isLoading) return variantPrimaryLoading;
      return variantPrimaryEnabled;
  }
};

export const ButtonWrapper = styled.button<ButtonWrapperProps>(props =>
  css({
    position: 'relative',
    /* To position the loading spinner correctly.
     * Pulled out of styles.ts so it works for the reset styles too.
     */
    ...buttonSizeStyles(props),
    ...buttonStateVariantStyles(props),
    width: props.fullWidth ? '100%' : 'auto',
  })
);

export const ButtonChildren = styled.span<ButtonChildrenProps>(props =>
  css({
    display: 'grid',
    gridAutoFlow: 'column',
    /*
     * Neat way to make sure children are spaced apart correctly
     * https://caniuse.com/#feat=multicolumn
     */
    columnGap: themeGet('space.space20')(props),
    justifyContent: 'center',
    alignItems: 'center',
    verticalAlign: 'middle',
    textDecoration: 'inherit',
    opacity: props.buttonState === 'loading' ? '0' : '100%',
  })
);

export const SpinnerWrapper = styled(Absolute)({
  display: 'flex',
  justifyContent: 'center',
  alignItems: 'center',
  lineHeight: '14px',
});
