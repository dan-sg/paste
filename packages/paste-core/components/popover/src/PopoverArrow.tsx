import * as React from 'react';
import {
  NonModalDialogArrowPrimitive,
  NonModalDialogArrowPrimitiveProps,
} from '@twilio-paste/non-modal-dialog-primitive';
import {useTheme} from '@twilio-paste/theme';

export type PopoverArrowProps = NonModalDialogArrowPrimitiveProps;

const PopoverArrow = React.forwardRef<HTMLDivElement, PopoverArrowProps>((props, ref) => {
  const theme = useTheme();

  return (
    <NonModalDialogArrowPrimitive
      {...props}
      size={theme.fontSizes.fontSize70}
      stroke={theme.borderColors.colorBorderLight}
      fill={theme.backgroundColors.colorBackgroundBody}
      ref={ref}
    />
  );
});
PopoverArrow.displayName = 'PopoverArrow';
export {PopoverArrow};
