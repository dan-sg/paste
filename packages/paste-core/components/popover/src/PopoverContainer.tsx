import * as React from 'react';
import {
  useNonModalDialogPrimitiveState,
  NonModalDialogPrimitivePopoverInitialState,
} from '@twilio-paste/non-modal-dialog-primitive';
import {PopoverContext} from './PopoverContext';

export type PopoverContainerProps = NonModalDialogPrimitivePopoverInitialState;

const PopoverContainer: React.FC<PopoverContainerProps> = ({children, placement, gutter, ...initialState}) => {
  const popover = useNonModalDialogPrimitiveState({placement, gutter, ...initialState});
  const value = React.useMemo(() => popover, Object.values(popover));
  const returnValue = <PopoverContext.Provider value={value}>{children}</PopoverContext.Provider>;
  return returnValue;
};

PopoverContainer.displayName = 'PopoverContainer';
export {PopoverContainer};
