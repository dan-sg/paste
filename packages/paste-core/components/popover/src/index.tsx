import * as React from 'react';
import {Box, BoxProps, safelySpreadBoxProps} from '@twilio-paste/box';
import {Button, ButtonProps} from '@twilio-paste/button';
import {
  useNonModalDialogPrimitiveState,
  NonModalDialogPrimitive,
  NonModalDialogPrimitiveProps,
  NonModalDialogDisclosurePrimitive,
  NonModalDialogDisclosurePrimitiveProps,
  NonModalDialogArrowPrimitive,
  NonModalDialogArrowPrimitiveProps,
} from '@twilio-paste/non-modal-dialog-primitive';
import {useTheme} from '@twilio-paste/theme';

export type PopoverProps = NonModalDialogPrimitiveProps & {'aria-label': string};
export type PopoverButtonProps = NonModalDialogDisclosurePrimitiveProps & ButtonProps;
export type PopoverArrowProps = NonModalDialogArrowPrimitiveProps;

const StyledPopover = React.forwardRef<HTMLDivElement, BoxProps>(({style, ...props}, ref) => {
  return (
    <Box
      {...safelySpreadBoxProps(props)}
      backgroundColor="colorBackgroundBody"
      borderStyle="solid"
      borderWidth="borderWidth10"
      borderColor="colorBorderLight"
      borderRadius="borderRadius20"
      boxShadow="shadowCard"
      maxWidth="size30"
      minWidth="size30"
      zIndex="zIndex10"
      _focus={{outline: 'none'}}
      style={style}
      ref={ref}
    />
  );
});

const Popover = React.forwardRef<HTMLDivElement, PopoverProps>((props, ref) => {
  return <NonModalDialogPrimitive {...props} as={StyledPopover} ref={ref} />;
});
Popover.displayName = 'Popover';
export {Popover};

const PopoverArrow = React.forwardRef<HTMLDivElement, PopoverArrowProps>((props, ref) => {
  const theme = useTheme();

  return (
    <NonModalDialogArrowPrimitive
      {...props}
      as="div"
      // @ts-ignore
      size="1.5rem"
      stroke={theme.borderColors.colorBorderLight}
      fill={theme.backgroundColors.colorBackgroundBody}
      ref={ref}
    />
  );
});
PopoverArrow.displayName = 'PopoverArrow';
export {PopoverArrow};

const PopoverButton = React.forwardRef<HTMLButtonElement, PopoverButtonProps>((props, ref) => {
  return (
    <NonModalDialogDisclosurePrimitive {...props} as={Button} ref={ref}>
      {props.children}
    </NonModalDialogDisclosurePrimitive>
  );
});
PopoverButton.displayName = 'PopoverButton';
export {PopoverButton};

export {useNonModalDialogPrimitiveState as usePopoverState};
