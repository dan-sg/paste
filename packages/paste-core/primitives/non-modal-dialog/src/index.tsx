export {
  usePopoverState as useNonModalDialogPrimitiveState,
  Popover as NonModalDialogPrimitive,
  PopoverProps as NonModalDialogPrimitiveProps,
  PopoverDisclosure as NonModalDialogDisclosurePrimitive,
  PopoverDisclosureProps as NonModalDialogDisclosurePrimitiveProps,
  PopoverArrow as NonModalDialogArrowPrimitive,
  PopoverArrowProps as NonModalDialogArrowPrimitiveProps,
} from '@twilio-paste/reakit-library';
