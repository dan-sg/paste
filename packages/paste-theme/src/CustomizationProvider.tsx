import * as React from 'react';
import styled from '@emotion/styled';
import {Global, css, SerializedStyles} from '@emotion/core';
import {themeGet} from '@styled-system/theme-get';
import {ThemeProvider} from 'emotion-theming';
import * as lodash from 'lodash';
import {DefaultTheme} from './themes';

interface EmotionStyledProps {
  theme: {};
}
const pasteGlobalStyles = (props: EmotionStyledProps): SerializedStyles => css`
  html {
    font-size: 100%;
  }
  body {
    margin: 0;
    font-size: ${themeGet('fontSizes.fontSize30')(props)};
  }
`;

export const pasteBaseStyles = (props: EmotionStyledProps): SerializedStyles => css`
  font-family: ${themeGet('fonts.fontFamilyText')(props)};
  font-size: ${themeGet('fontSizes.fontSize30')(props)};
  line-height: ${themeGet('lineHeights.lineHeight30')(props)};
  color: ${themeGet('textColors.colorText')(props)};
  font-weight: ${themeGet('fontWeights.fontWeightNormal')(props)};
  *,
  *::after,
  *::before {
    box-sizing: border-box;
  }
`;

export const StyledBase = styled.div`
  ${pasteBaseStyles}
`;

export interface CustomizationProviderProps {
  customBreakpoints?: string[];
  theme?: {};
  elements?: {};
}

const CustomizationProvider: React.FunctionComponent<CustomizationProviderProps> = ({
  customBreakpoints,
  theme,
  elements,
  ...props
}) => {
  const breakpoints = customBreakpoints || DefaultTheme.breakpoints;
  const baseTheme = lodash.cloneDeep(DefaultTheme);

  const themeObject = lodash.merge(baseTheme, theme);
  const providerThemeProp = React.useMemo(
    () => ({
      ...themeObject,
      breakpoints,
      CUSTOMIZATION_OPT_IN_OVERRIDE_DO_NOT_USE: true,
      elements,
    }),
    [themeObject, breakpoints, elements]
  );

  return (
    <ThemeProvider theme={providerThemeProp}>
      <Global styles={pasteGlobalStyles({theme: providerThemeProp})} />
      <StyledBase {...props} />
    </ThemeProvider>
  );
};

export {CustomizationProvider};
