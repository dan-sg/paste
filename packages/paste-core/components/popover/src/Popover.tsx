import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Box, BoxProps, safelySpreadBoxProps} from '@twilio-paste/box';
import {NonModalDialogPrimitive} from '@twilio-paste/non-modal-dialog-primitive';
import {PopoverArrow} from './PopoverArrow';
import {PopoverContext} from './PopoverContext';

export interface PopoverProps {
  'aria-label': string;
  children: React.ReactNode;
  tabIndex?: number;
}

const StyledPopover = React.forwardRef<HTMLDivElement, BoxProps>(({style, ...props}, ref) => {
  return (
    <Box
      {...safelySpreadBoxProps(props)}
      backgroundColor="colorBackgroundBody"
      borderStyle="solid"
      borderWidth="borderWidth10"
      borderColor="colorBorderLight"
      borderRadius="borderRadius20"
      boxShadow="shadowCard"
      maxWidth="size30"
      minWidth="size30"
      zIndex="zIndex10"
      _focus={{outline: 'none'}}
      style={style}
      ref={ref}
    />
  );
});

const Popover = React.forwardRef<HTMLDivElement, PopoverProps>(({children, ...props}, ref) => {
  const popover = React.useContext(PopoverContext);
  return (
    <NonModalDialogPrimitive {...(popover as any)} {...props} as={StyledPopover} ref={ref}>
      <PopoverArrow {...(popover as any)} />
      {children}
    </NonModalDialogPrimitive>
  );
});

if (process.env.NODE_ENV === 'development') {
  Popover.propTypes = {
    'aria-label': PropTypes.string.isRequired,
    children: PropTypes.node.isRequired,
    tabIndex: PropTypes.number,
  };
}

Popover.displayName = 'Popover';
export {Popover};
