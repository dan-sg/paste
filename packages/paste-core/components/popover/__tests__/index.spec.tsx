import * as React from 'react';
import {axe} from 'jest-axe';
import {render, screen} from '@testing-library/react';
import {Theme} from '@twilio-paste/theme';
import {usePopoverState, Popover, PopoverArrow, PopoverButton} from '../src';

const PopoverMock: React.FC<{}> = () => {
  const popover = usePopoverState({baseId: 'popover-example'});
  return (
    <Theme.Provider theme="console">
      <PopoverButton {...popover} variant="primary">
        Open popover
      </PopoverButton>
      <Popover {...popover} tabIndex={0} aria-label="Popover">
        <PopoverArrow {...popover} />
        This is the Twilio styled popover that you can use in all your applications.
      </Popover>
    </Theme.Provider>
  );
};

const VisiblePopoverMock: React.FC<{}> = () => {
  const popover = usePopoverState({baseId: 'popover-example', visible: true});
  return (
    <Theme.Provider theme="console">
      <PopoverButton {...popover} variant="primary">
        Open popover
      </PopoverButton>
      <Popover {...popover} tabIndex={0} aria-label="Popover">
        <PopoverArrow {...popover} />
        This is the Twilio styled popover that you can use in all your applications.
      </Popover>
    </Theme.Provider>
  );
};

describe('Popover', () => {
  describe('Render', () => {
    it('should render', () => {
      const {asFragment} = render(<PopoverMock />);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should render a popover button with aria attributes', () => {
      render(<PopoverMock />);
      const renderedPopoverButton = screen.getByRole('button');
      expect(renderedPopoverButton.getAttribute('aria-haspopup')).toEqual('dialog');
      expect(renderedPopoverButton.getAttribute('aria-controls')).toEqual('popover-example');
      expect(renderedPopoverButton.getAttribute('aria-expanded')).toEqual('false');
    });

    it('should render a popover button with expanded aria attributes when popover is visible', () => {
      render(<VisiblePopoverMock />);
      const renderedPopoverButton = screen.getByRole('button');
      expect(renderedPopoverButton.getAttribute('aria-expanded')).toEqual('true');
    });

    it('should render a popover', () => {
      render(<PopoverMock />);
      const renderedPopover = screen.getByLabelText('Popover');
      expect(renderedPopover.getAttribute('role')).toEqual('dialog');
    });
  });

  describe('Accessibility', () => {
    it('Should have no accessibility violations', async () => {
      const {container} = render(<PopoverMock />);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });
});
