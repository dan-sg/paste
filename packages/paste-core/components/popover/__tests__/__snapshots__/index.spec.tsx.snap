// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover Render should render 1`] = `
<DocumentFragment>
  .emotion-3 {
  color: rgb(40,42,43);
  font-size: 0.875rem;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  line-height: 1.5rem;
  font-weight: 400;
}

.emotion-1 {
  position: relative;
  padding: 0.5rem 1rem;
  border-radius: 4px;
  font-size: 0.875rem;
  line-height: 1.5rem;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  border: none;
  display: inline-block;
  outline: none;
  background: none;
  -webkit-transition: background-color 100ms ease-in,box-shadow 100ms ease-in;
  transition: background-color 100ms ease-in,box-shadow 100ms ease-in;
  font-family: 'Whitney SSm A','Whitney SSm B','Helvetica Neue',Helvetica,Arial,sans-serif;
  font-weight: 500;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  color: rgb(255,255,255);
  background-color: rgb(2,122,197);
  box-shadow: inset 0 0 0 0.125rem rgb(2,122,197);
  width: auto;
}

.emotion-1::-moz-focus-inner {
  border: none;
}

.emotion-1:focus,
.emotion-1:active {
  box-shadow: 0 0 0 4px rgba(2,122,197,0.8);
}

.emotion-1:hover,
.emotion-1:focus,
.emotion-1:active {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-1:hover,
.emotion-1:focus,
.emotion-1:active {
  color: rgb(255,255,255);
}

.emotion-1:hover {
  background-color: rgb(3,44,94);
  box-shadow: inset 0 0 0 0.125rem rgb(3,44,94);
}

.emotion-1:focus {
  background-color: rgb(2,122,197);
  box-shadow: inset 0 0 0 0.125rem rgb(3,44,94), 0 0 0 4px rgba(2,122,197,0.8);
}

.emotion-1:active {
  background-color: rgb(0,62,130);
  box-shadow: inset 0 0 0 0.125rem rgb(3,44,94), 0 0 0 4px rgba(2,122,197,0.8);
}

.emotion-0 {
  display: grid;
  grid-auto-flow: column;
  -webkit-column-gap: 0.25rem;
  column-gap: 0.25rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  vertical-align: middle;
  -webkit-text-decoration: inherit;
  text-decoration: inherit;
  opacity: 100%;
}

.emotion-2 {
  box-sizing: border-box;
  background-color: rgb(255,255,255);
  border-style: solid;
  border-width: 1px;
  border-color: rgb(228,231,233);
  border-radius: 4px;
  box-shadow: 0 2px 4px 0 rgba(40,42,43,0.3);
  max-width: 18.5rem;
  min-width: 18.5rem;
  z-index: 10;
}

.emotion-2:focus {
  outline: none;
}

<div
    class="emotion-3"
  >
    <button
      aria-busy="false"
      aria-controls="popover-example"
      aria-expanded="false"
      aria-haspopup="dialog"
      class="emotion-1"
      type="button"
    >
      <span
        class="emotion-0"
      >
        Open popover
      </span>
    </button>
    <div
      aria-label="Popover"
      class="emotion-2"
      data-dialog="true"
      hidden=""
      id="popover-example"
      role="dialog"
      style="display: none;"
      tabindex="0"
    >
      <div
        style="font-size: 1.5rem; width: 1em; height: 1em; pointer-events: none; bottom: 100%;"
      >
        <svg
          style="transform: rotateZ(360deg);"
          viewBox="0 0 30 30"
        >
          <path
            d="M23.7,27.1L17,19.9C16.5,19.3,15.8,19,15,19s-1.6,0.3-2.1,0.9l-6.6,7.2C5.3,28.1,3.4,29,2,29h26
        C26.7,29,24.6,28.1,23.7,27.1z"
            fill="rgb(255, 255, 255)"
            stroke="rgb(228, 231, 233)"
          />
          <path
            d="M23,27.8c1.1,1.2,3.4,2.2,5,2.2h2H0h2c1.7,0,3.9-1,5-2.2l6.6-7.2c0.7-0.8,2-0.8,2.7,0L23,27.8L23,27.8z"
            fill="rgb(255, 255, 255)"
          />
        </svg>
      </div>
      This is the Twilio styled popover that you can use in all your applications.
    </div>
  </div>
</DocumentFragment>
`;
